synapse/push/__init__.py

```diff
@@ -266,23 +281,64 @@ def dispatch_push(self, p, tweaks):
         """
         pass
 
-    def reset_badge_count(self):
-        pass
+    @defer.inlineCallbacks
+    def update_badge(self):
+        new_badge = yield self._get_badge_count()
+        if self.badge != new_badge:
+            self.badge = new_badge
+            yield self.send_badge(self.badge)
 
-    def presence_changed(self, state):
+    def send_badge(self, badge):
         """
-        We clear badge counts whenever a user's last_active time is bumped
-        This is by no means perfect but I think it's the best we can do
-        without read receipts.
+        Overridden by implementing classes to send an updated badge count
         """
-        if 'last_active' in state.state:
-            last_active = state.state['last_active']
-            if last_active > self.last_last_active_time:
-                self.last_last_active_time = last_active
-                if self.has_unread:
-                    logger.info("Resetting badge count for %s", self.user_id)
-                    self.reset_badge_count()
-                    self.has_unread = False
+        pass
+
+    @defer.inlineCallbacks
+    def _get_badge_count(self):
+        membership_list = (Membership.INVITE, Membership.JOIN)
+
+        room_list = yield self.store.get_rooms_for_user_where_membership_is(
+            user_id=self.user_id,
+            membership_list=membership_list
+        )
+
+        user_is_guest = yield self.store.is_guest(self.user_id)
+
+        # XXX: importing inside method to break circular dependency.
+        # should sort out the mess by moving all this logic out of
+        # push/__init__.py and probably moving the logic we use from the sync
+        # handler to somewhere more amenable to re-use.
+        from synapse.handlers.sync import SyncConfig
+        sync_config = SyncConfig(
+            user=UserID.from_string(self.user_id),
+            filter=FilterCollection({}),
+            is_guest=user_is_guest,
+        )
+        now_token = yield self.hs.get_event_sources().get_current_token()
+        sync_handler = self.hs.get_handlers().sync_handler
+        _, ephemeral_by_room = yield sync_handler.ephemeral_by_room(
+            sync_config, now_token
+        )
```

Comment:
Rather than abusing SyncConfig it'd probably be cleaner to add a storage function `get_receipts_for_room`, e.g.:

```python
    @cached(num_args=3)
    def get_receipts_for_room_for_room(self, room_id, user_id, receipt_type):
        return self._simple_select_list(
            table="receipts_linearized",
            keyvalues={
                "room_id": room_id,
                "receipt_type": receipt_type,
                "user_id": user_id,
            },
            retcols=("event_id"),
            desc="get_receipts_for_room",
        )
```

or possibly using `_simple_select_one_onecol`